### PR TITLE
Don't create published solution rep token when unpublished

### PR DESCRIPTION
**Pull request by ErikSchierboom**

### CODE REVIEW

**Review by iHiD — APPROVED** [review]
Thanks :)

File: app/models/user/reputation_tokens/published_solution_token.rb [hunk 1]
```diff
@@ -9,6 +9,8 @@ class User::ReputationTokens::PublishedSolutionToken < User::ReputationToken
     self.track = solution.track unless track
     self.exercise = solution.exercise unless exercise
     self.earned_on = solution.published_at unless earned_on
+
+    raise NuPublishedSolutionForReputationTokenError unless earned_on
```
Comment (iHiD) [inline app/models/user/reputation_tokens/published_solution_token.rb]:
```suggestion
    raise NoPublishedSolutionForReputationTokenError unless earned_on.present?
```

File: config/initializers/exceptions.rb [hunk 2]
```diff
@@ -72,3 +72,5 @@ class RequestLockHasExpired < RuntimeError; end
 class InvalidDiscussionStatusError < RuntimeError; end
 
 class TrainingDataCodeTagsSampleLockedError < RuntimeError; end
+
+class NuPublishedSolutionForReputationTokenError < RuntimeError; end
```
Comment (iHiD) [inline config/initializers/exceptions.rb]:
```suggestion
class NoPublishedSolutionForReputationTokenError < RuntimeError; end
```

File: app/commands/user/reputation_token/create.rb [hunk 3]
```diff
@@ -26,5 +26,7 @@ def call
     rescue ActiveRecord::RecordNotUnique
       return klass.find_by!(user:, uniqueness_key: token.uniqueness_key)
     end
+  rescue NuPublishedSolutionForReputationTokenError
```
Comment (ErikSchierboom) [inline app/commands/user/reputation_token/create.rb]:
```suggestion
  rescue NoPublishedSolutionForReputationTokenError
```

**Pull request merged**